handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -79,6 +83,42 @@ protected ApplicationProtocolNegotiationHandler(String fallbackProtocol) {
         this.fallbackProtocol = ObjectUtil.checkNotNull(fallbackProtocol, "fallbackProtocol");
     }
 
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        drainBacklog();
+        backlogList.recycle();
+        super.handlerRemoved(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        backlogList.add(msg);
+    }
+
+    /**
+     * Release all backlog into pipeline from List.
+     */
+    private void releaseBacklog() {
+        for (Object o : backlogList) {
+            ctx.fireChannelRead(o);
+        }
+    }
+
+    /**
+     * Drain (or discard) all backlog from List
+     */
+    private void drainBacklog() {
```

Comment:
imho we should name this `releaseBacklog` as we release all the buffers... Also you should call `clear()` to ensure we not try to process the same stuff again